capsules/src/log_storage.rs

```diff
@@ -0,0 +1,936 @@
+//!         )
+//!     );
+//!     kernel::hil::flash::HasClient::set_client(&sam4l::flashcalw::FLASH_CONTROLLER, log_storage);
+//!     log_storage.initialize_callback_handle(dynamic_deferred_caller.register(log_storage).expect("no deferred call slot available for log storage"));
+//!
+//!     log_storage.set_read_client(log_storage_read_client);
+//!     log_storage.set_append_client(log_storage_append_client);
+//! ```
+
+use crate::storage_interface::{
+    LogRead, LogReadClient, LogWrite, LogWriteClient, OperationResult, StorageCookie, StorageLen,
+};
+use core::cell::Cell;
+use core::convert::TryFrom;
+use core::mem::size_of;
+use core::unreachable;
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::common::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use kernel::hil::flash::{self, Flash};
+use kernel::ReturnCode;
+
+/// Maximum page header size.
+pub const PAGE_HEADER_SIZE: usize = size_of::<usize>();
+/// Maximum entry header size.
+pub const ENTRY_HEADER_SIZE: usize = size_of::<usize>();
+
+/// Byte used to pad the end of a page.
+const PAD_BYTE: u8 = 0xFF;
+
+/// Log state keeps track of any in-progress asynchronous operations.
+#[derive(Clone, Copy, PartialEq)]
+enum State {
+    Idle,
+    Read,
+    Seek,
+    Append,
+    Sync,
+    Erase,
+    Invalid,
+}
+
+pub struct LogStorage<'a, F: Flash + 'static, RC: LogReadClient, WC: LogWriteClient> {
+    /// Underlying storage volume.
+    volume: &'static [u8],
+    /// Capacity of log in bytes.
+    capacity: StorageLen,
+    /// Flash interface.
+    driver: &'a F,
+    /// Buffer for a flash page.
+    pagebuffer: TakeCell<'static, F::Page>,
+    /// Size of a flash page.
+    page_size: usize,
+    /// Whether or not the log is circular.
+    circular: bool,
+    /// Read client using LogStorage.
+    read_client: OptionalCell<&'a RC>,
+    /// Append client using LogStorage.
+    append_client: OptionalCell<&'a WC>,
+
+    /// Current operation being executed, if asynchronous.
+    state: Cell<State>,
+    /// Cookie within log to read from.
+    read_cookie: Cell<usize>,
+    /// Cookie within log to append to.
+    append_cookie: Cell<usize>,
+    /// Oldest cookie still in log.
+    oldest_cookie: Cell<usize>,
+
+    /// Deferred caller for deferring client callbacks.
+    deferred_caller: &'a DynamicDeferredCall,
+    /// Handle for deferred caller.
+    handle: OptionalCell<DeferredCallHandle>,
+
+    // Note: for saving state across stack ripping.
+    /// Client-provided buffer to write from.
+    buffer: TakeCell<'static, [u8]>,
+    /// Length of data within buffer.
+    length: Cell<StorageLen>,
+    /// Whether or not records were lost in the previous append.
+    records_lost: Cell<bool>,
+    /// Error returned by previously executed operation (or SUCCESS).
+    error: Cell<ReturnCode>,
+}
+
+impl<'a, F: Flash + 'static, RC: LogReadClient, WC: LogWriteClient> LogStorage<'a, F, RC, WC> {
```

Comment:
Does this need to be parameterized by the LogReadClient and LogWriteClient types, or could they be dyn?

Reply:
Oh yeah, using dyn is much cleaner, thanks.